### ISSUE TITLE
Feature Request - ARM64 support

### ISSUE THREAD

**Issue by NeomMob** [cefsharp/CefSharp#2944]
We are starting now to see some new devices from Microsoft using ARM64 architecture. I am wondering if there is a forecasted support for them.

**Comment by kpreisser** [comment]
See also dotnet/winforms#2053 for the progress of ARM64 support in .NET Core, including WinForms and WPF.

**Comment by amaitland** [comment]
I personally have no plans to support `Win ARM64`, I don't own a suitable device.

For background.
- There are no builds on http://opensource.spotify.com/cefbuilds/index.html not clear if there will be any publicly available. 
- From what I can tell `CEF` support (probably `Chomium` in general) for building `Win ARM64` is still in it's early stages.
  - https://bitbucket.org/chromiumembedded/cef/issues/2773/windows-add-arm64-build-support
  - `https://bitbucket.org/chromiumembedded/cef/pull-requests/266
- I'm not sure `Chrome` has actually released an official version, so it's likely too early to look at from a stability point of view.

**Comment by amaitland** [comment]
Someone else is welcome to take responsibility for adding and supporting `Win ARM64`. An ongoing commitment would be required.

**Comment by amaitland** [comment]
According to https://bitbucket.org/chromiumembedded/cef/issues/1990/linux-add-arm-build-support#comment-55787851 ARM64 builds will be available on the Spotify build server.

**Comment by amaitland** [comment]
> Official Windows ARM64 builds starting with M88 will be available shortly from https://cef-builds.spotifycdn.com/index.html#windowsarm64.

As per https://bitbucket.org/chromiumembedded/cef/issues/2773/windows-add-arm64-build-support#comment-59706180

This is blocked on the linker errors when attempting to compiler `CefSharp.Core.Runtime` targeting `.Net 5.0` (required for `ARM64`). As reported in  https://github.com/cefsharp/CefSharp/issues/3284#issuecomment-734779646 which I've confirmed as being a problem.

```
2>MSVCMRTD_netcore.LIB(mstartup.obj) : error LNK2022: metadata operation failed (80131195) :
2>MSVCMRTD_netcore.LIB(mstartup.obj) : error LNK2022: metadata operation failed (80131195) :
2>MSVCMRTD_netcore.LIB(mstartup.obj) : error LNK2022: metadata operation failed (80131195) :
2>LINK : fatal error LNK1255: link failed because of metadata errors
```

The `CefSharp.BrowserSubprocess.Core` project compiles successfully, it's less complicated though, it's possible there's just a linker setting that's causing the problem or we might need to raise an issue with `Net 5 Team`.

**Comment by amaitland** [comment]
For anyone interested in tacking this it looks like the first build is up on https://cef-builds.spotifycdn.com/index.html#windowsarm64

First step would be working out how to generate `libcef_dll_wrapper.lib`

Currently https://github.com/cefsharp/cef-binary/blob/master/build.ps1 downloads, builds and packages the `x86/x64` versions, it should be possible (haven't tried obviously) to extend this to download, build and package the `arm64` target binaries.

**Comment by kpreisser** [comment]
> This is blocked on the linker errors when attempting to compiler `CefSharp.Core.Runtime` targeting `.Net 5.0` (required for `ARM64`). As reported in [#3284 (comment)](https://github.com/cefsharp/CefSharp/issues/3284#issuecomment-734779646) which I've confirmed as being a problem.

Note that switching to `net5.0` might not actually be necessary for supporting `arm64` in CefSharp. With `cef.sdk 88.1.2` that includes the ARM64 versions of `libcef.lib` and `libcef_dll_wrapper.lib` (from cefsharp/cef-binary#90), I was able to successfully compile `CefSharp.Core.Runtime` and `CefSharp.BrowseSubprocess.Core` after adding `arm64` as platform target, and confirmed with `dumpbin.exe /headers` that the produced DLLs (`ijwhost.dll`, `CefSharp.Core.Runtime.dll`, `CefSharp.BrowserSubprocess.Core.dll`) are actually ARM64 DLLs.

I still need to check whether building a WinForms/OffScreen app for `win-arm64` (with nightly version of NET 6.0, since .NET 5.0 doesn't yet include the Windows Desktop components for ARM64) and running it on a Windows 10 on ARM machine actually works.

**Comment by amaitland** [comment]
It looks like `.Net 5.0.8` now officially supports targeting `WinForms/WPF` for `ARM64.

https://devblogs.microsoft.com/dotnet/net-july-2021/#windows-arm64

Once `AppVeyor` has updated to include the newest `SDK` we can look at officially releasing `ARM64` builds.

https://www.appveyor.com/docs/windows-images-software/#net-framework

**Comment by A-Ovchinnikov-mx** [comment]
Hi @amaitland It looks like appveyor has updated the .NET SDK to 5.0.302 which is the version that includes ARM64 SDK. How complicated would it be to build CEFSharp for Windows ARM64 now? Thanks!

**Comment by amaitland** [comment]
You can test build [92.0.250-CI4152](https://ci.appveyor.com/project/cefsharp/cefsharp/builds/40378359) which is now on 
https://www.myget.org/gallery/cefsharp

It was hopefully build with the latest `SDK`.

**Comment by amaitland** [comment]
> I also wanted give an updated feedback from our side:

@kpreisser Great, thanks for the feedback 👍 

> But I suppose you could just go ahead and give it to the public since it doesn't jeopardize other architectures in any way,

@A-Ovchinnikov-mx I would personally like to include `ARM64` support in version `93` as it would save me some time as I can finally create the `release` branch from `master` which includes `ARM64` support.

**Comment by amaitland** [comment]
Anyone who is interested in `ARM64` support and has some time please  check out a `CI` build and report back. Thanks.

**Comment by amaitland** [comment]
`ARM64` has been included in `93.1.110-pre` see https://github.com/cefsharp/CefSharp/issues/3780#issuecomment-913087509 for release details.

Please be aware that I'm relying on community to provide support for `ARM64` as I personally don't own an `ARM64` device.

**Comment by kpreisser** [comment]
Additionally, I think we need to change `DependencyChecker` to not check for `D3DCompiler_47.dll` for `win-arm64` (as it's not included in `chromiumembeddedframework.runtime.win-arm64`):
https://github.com/cefsharp/CefSharp/blob/15cf58c273c90107745056e02158b3f9d7eba55d/CefSharp/DependencyChecker.cs#L57-L66

Currently, it always checks for the presence of `D3DCompiler_47.dll`. This works currently due to the stub DLL deployed by the .NET SDK, but once this is fixed and that file is no longer copied, the dependency check would fail when running on `ARM64`.

**Comment by kpreisser** [comment]
> I have reported this issue about `D3DCompiler_47.dll` in [dotnet/wpf#5462](https://github.com/dotnet/wpf/issues/5462).

Just a follow-up: The fix has been backported to .NET 6.0 and is expected to be included in the June 2022 release (6.0.6).
